README.md
# embedded-redis [![Build Status](https://travis-ci.org/fmonniot/embedded-redis.svg?branch=master)](https://travis-ci.org/fmonniot/embedded-redis) [ ![Download](https://api.bintray.com/packages/fmonniot/maven/embedded-redis/images/download.svg) ](https://bintray.com/fmonniot/maven/embedded-redis/_latestVersion)

Redis embedded server for Java integration testing

_This is a fork of https://github.com/kstyrc/embedded-redis and remove support for several features:_
* _Embedded redis binaries, you MUST provide them yourself_
* _Java 6 support_
* _Spring data support (should work but not tested anymore)_

_It also add support for some other features:_
* _Support for Redis Cluster_
* _Cleanup the API_
* _Let user access Redis instances logs_
* _More consistent use of PortProvider interface_

# Maven dependency

This library is available on jCenter:
```xml
<dependency>
  <groupId>eu.monniot.redis</groupId>
  <artifactId>embedded-redis</artifactId>
  <version>1.6.0</version>
</dependency>
```

# Usage

## Redis Server

Running RedisServer is as simple as:
```java
RedisServer redisServer = new RedisServer(6379);
redisServer.start();
// do some work
redisServer.stop();
```

This works as long as you have a `redis-server-3.0.7-*` executables in your path (see
[redis/embedded/RedisExecProvider.java](https://github.com/fmonniot/embedded-redis/blob/master/src/main/java/redis/embedded/RedisExecProvider.java#L26)
for the complete list of default binaries paths).

You can also provide RedisServer with your own executable:
```java
// Give an OS-independent matrix
RedisExecProvider provider = RedisExecProvider.build()
  .override(OS.UNIX, "/path/to/unix/redis")
  .override(OS.WINDOWS, Architecture.x86, "/path/to/windows/redis")
  .override(OS.Windows, Architecture.x86_64, "/path/to/windows/redis")
  .override(OS.MAC_OS_X, Architecture.x86, "/path/to/macosx/redis")
  .override(OS.MAC_OS_X, Architecture.x86_64, "/path/to/macosx/redis")
  
RedisServer redisServer = new RedisServer(provider, 6379);
```

You can also use fluent API to create RedisServer:
```java
RedisServer redisServer = new RedisServer.Builder()
  .redisExecProvider(customRedisProvider)
  .port(6379)
  .slaveOf("locahost", 6378)
  .configFile("/path/to/your/redis.conf")
  .build();
```

Or even create the redis.conf file from the builder:
```java
RedisServer redisServer = new RedisServer.Builder()
  .redisExecProvider(customRedisProvider)
  .port(6379)
  .slaveOf("locahost", 6378)
  .setting("daemonize no")
  .setting("appendonly no")
  .build();
```

The `JedisUtil` class contains utility methods to get the list of port in a Jedis friendly format.

## Setting up a cluster

Embedded Redis has support for [Redis Cluster](http://redis.io/topics/cluster-tutorial).

```java
Redis cluster = new RedisCluster.Builder()
        .withServerBuilder(myOwnRedisServerBuilder)
        .serverPorts(Arrays.asList(42000,42001,42002,42003,42004,42005))
        .numOfReplicates(1)
        .numOfRetries(42)
        .build()

cluster.start()
cluster.stop()
```

## Setting up a sentinel cluster

Embedded Redis has support for HA Redis clusters with Sentinels and master-slave replication

#### Using ephemeral ports

A simple Redis cluster on ephemeral ports, with setup similar to that from production, would look like this:


```java
//creates a cluster with 3 sentinels, quorum size of 2 and 3 replication groups, each with one master and one slave
cluster = new SentinelCluster.Builder()
                .ephemeral()
                .sentinelCount(3)
                .quorumSize(2)
                .replicationGroup("master1", 1)
                .replicationGroup("master2", 1)
                .replicationGroup("master3", 1)
                .build();
cluster.start();
cluster.stop();
```

The above example starts Redis cluster on ephemeral ports, which you can later get with ```cluster.ports()```,
which will return a list of all ports of the cluster. You can also get ports of sentinels with ```cluster.sentinelPorts()```
or servers with ```cluster.serverPorts()```. 

This library also includes an utility for helping you integrating it with Jedis
 
 ```java
 //retrieve ports on which sentinels have been started, using a simple Jedis utility class
 Set<String> jedisSentinelHosts = JedisUtil.sentinelHosts(cluster);
 // testing code that requires redis running
 JedisSentinelPool pool = new JedisSentinelPool("master1", jedisSentinelHosts);
 ```

#### Using predefined ports
You can also start Redis cluster on predefined ports and even mix both approaches:

```java
final List<Integer> sentinels = Arrays.asList(26739, 26912);
final List<Integer> group1 = Arrays.asList(6667, 6668);
final List<Integer> group2 = Arrays.asList(6387, 6379);

//creates a cluster with 3 sentinels, quorum size of 2 and 3 replication groups, each with one master and one slave
cluster = new SentinelCluster.Builder()
                .sentinelPorts(sentinels)
                .quorumSize(2)
                .serverPorts(group1).replicationGroup("master1", 1)
                .serverPorts(group2).replicationGroup("master2", 1)
                .ephemeralServers().replicationGroup("master3", 1)
                .build();
cluster.start();
```

The above will create and start a cluster with sentinels on ports ```26739, 26912```, first replication group on ```6667, 6668```,
second replication group on ```6387, 6379``` and third replication group on ephemeral ports.

## Redis version

Since this library doesn't provides binary by default, and thanks to @xhoong, we can now further customize redis executables
and the pattern used to detect when they became available. This is fairly useful when using a version of redis not supported
by this library:

```java
RedisExecProvider provider = RedisExecProvider.build()
  .override(OS.UNIX, Architecture.x86_64, RedisExecutable.build("redis-4.2-bin", "Ready to accept connections"))
```

# License

Licensed under the Apache License, Version 2.0


# Contributors

 * François Monniot ([@fmonniot](https://github.com/fmonniot))
 * Krzysztof Styrc ([@kstyrc](https://github.com/kstyrc))
 * Piotr Turek ([@turu](https://github.com/turu))
 * anthonyu ([@anthonyu](https://github.com/anthonyu))
 * Artem Orobets ([@enisher](https://github.com/enisher))
 * Sean Simonsen ([@SeanSimonsen](https://github.com/SeanSimonsen))
 * Rob Winch ([@rwinch](https://github.com/rwinch))
 * Pavel Moukhataev ([@mpashka](https://github.com/mpashka))
 * Xavier L. ([@khoong](https://github.com/xhoong))


CHANGELOG.md
This project is using GitHub releases from version 1.4.0 onward for changes.

## 1.4.0

 * Provides a `PortProvider#copy()` method to make a new copy of the `PortProvider` based on its constructor arguments (it will NOT copy the current state of the iterator) 

### 1.3.1

 * Fix an immutability problem in `RedisServer.Builder#copy()`
 * Provides a `RedisServer.Builder#settings()` method to preview the settings of a Builder

### 1.3.0

 * Provides a `RedisServer.Builder#copy()` method
 * `RedisCluster.Builder` do not `reset()` a provided `RedisServer.Builder` but copy it before applying additional config

### 1.2.2

 * Fix a possible class path loading error

### 1.2.1

 * Fix a possible deadlock in the AbstractRedisInstance.errors()

### 1.2.0

 * Use simpler errors() method in Redis interface

### 1.1.0

 * Remove support for RedisServer(File, int) constructor
 * Remove redis binaries from distributed jar library
 * Improve PortProvider interface and implementation
 * RedisCluster.Builder can now use a PortProvider instead of a collection of port

### 1.0.3

 * Deprecates use of `builder()` methods, use `Builder` class directly

### 1.0.1 -> 1.0.2
 * Miscellaneous fixes to get this lib on jCenter 

### 1.0
 * Rename RedisCluster into SentinelCluster
 * Introduce a real redis cluster support (thx to @3Dragan for the initial work)
 * Make all Builder inner classes
 * Possibility to provide your own OutputStream for logs (instead of System.out)

### 0.6
 * Support JDK 6 +

### 0.5
 * OS detection fix
 * redis binary per OS/arch pair
 * Updated to 2.8.19 binary for Windows

### 0.4 
 * Updated for Java 8
 * Added Sentinel support
 * Ability to create arbitrary clusters on arbitrary (ephemeral) ports
 * Updated to latest guava 
 * Throw an exception if redis has not been started
 * Redis errorStream logged to System.out

### 0.3
 * Fluent API for RedisServer creation

### 0.2
 * Initial decent release


src/main/java/redis/embedded/RedisSentinel.java
package redis.embedded;

import com.google.common.base.Preconditions;
import com.google.common.io.Files;
import redis.embedded.exceptions.RedisBuildingException;

import java.io.File;
import java.io.IOException;
import java.nio.charset.Charset;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Pattern;

public class RedisSentinel extends AbstractRedisInstance {
    private Pattern redisReadyPat = RedisExecProvider.DEFAULT_REDIS_READY_PATTERN;

    public RedisSentinel(List<String> args, int port) {
        super(port);
        this.args = new ArrayList<>(args);
    }

    void setRedisReadyPat(Pattern redisReadyPat) {
        Preconditions.checkNotNull(redisReadyPat);
        this.redisReadyPat = redisReadyPat;
    }

    @Deprecated
    public static Builder builder() {
        return new Builder();
    }

    @Override
    protected Pattern redisReadyPattern() {
        return redisReadyPat;
    }

    @SuppressWarnings("unused")
    public static class Builder {
        private static final String LINE_SEPARATOR = System.getProperty("line.separator");
        private static final String CONF_FILENAME = "embedded-redis-sentinel";
        private static final String MASTER_MONITOR_LINE = "sentinel monitor %s 127.0.0.1 %d %d";
        private static final String DOWN_AFTER_LINE = "sentinel down-after-milliseconds %s %d";
        private static final String FAIL_OVER_LINE = "sentinel failover-timeout %s %d";
        private static final String PARALLEL_SYNCS_LINE = "sentinel parallel-syncs %s %d";
        private static final String PORT_LINE = "port %d";
        private static final long DEFAULT_STARTUP_TIMEOUT_MS = 3000;

        private File executable;
        private RedisExecProvider redisExecProvider = RedisExecProvider.defaultProvider();
        private Integer port = 26379;
        private int masterPort = 6379;
        private String masterName = "myMaster";
        private long downAfterMilliseconds = 60000L;
        private long failOverTimeout = 180000L;
        private int parallelSyncs = 1;
        private int quorumSize = 1;
        private String sentinelConf;

        private StringBuilder redisConfigBuilder;
        private boolean logProcessOutput;
        private long startupTimeoutMs = DEFAULT_STARTUP_TIMEOUT_MS;

        public Builder redisExecProvider(RedisExecProvider redisExecProvider) {
            this.redisExecProvider = redisExecProvider;
            return this;
        }

        public Builder port(Integer port) {
            this.port = port;
            return this;
        }

        public Builder masterPort(Integer masterPort) {
            this.masterPort = masterPort;
            return this;
        }

        public Builder masterName(String masterName) {
            this.masterName = masterName;
            return this;
        }

        public Builder quorumSize(int quorumSize) {
            this.quorumSize = quorumSize;
            return this;
        }

        public Builder downAfterMilliseconds(Long downAfterMilliseconds) {
            this.downAfterMilliseconds = downAfterMilliseconds;
            return this;
        }

        public Builder failOverTimeout(Long failoverTimeout) {
            this.failOverTimeout = failoverTimeout;
            return this;
        }

        public Builder parallelSyncs(int parallelSyncs) {
            this.parallelSyncs = parallelSyncs;
            return this;
        }

        public Builder configFile(String redisConf) {
            if (redisConfigBuilder != null) {
                throw new RedisBuildingException("Redis configuration is already partially build using setting(String) method!");
            }
            this.sentinelConf = redisConf;
            return this;
        }

        @SuppressWarnings("Duplicates")
        public Builder setting(String configLine) {
            if (sentinelConf != null) {
                throw new RedisBuildingException("Redis configuration is already set using redis conf file!");
            }

            if (redisConfigBuilder == null) {
                redisConfigBuilder = new StringBuilder();
            }

            redisConfigBuilder.append(configLine);
            redisConfigBuilder.append(LINE_SEPARATOR);
            return this;
        }

        public Builder logProcessOutput() {
            this.logProcessOutput = true;
            return this;
        }

        public Builder startupTimeoutMs(long startupTimeoutMs) {
            this.startupTimeoutMs = startupTimeoutMs;
            return this;
        }

        public RedisSentinel build() {
            tryResolveConfAndExec();
            List<String> args = buildCommandArgs();
            RedisSentinel redisSentinel = new RedisSentinel(args, port);
            redisSentinel.setLogProcessOutput(logProcessOutput);
            redisSentinel.setStartupTimeoutMs(startupTimeoutMs);
            redisSentinel.setRedisReadyPat(redisExecProvider.getExecutableStartPattern());
            return redisSentinel;
        }

        private void tryResolveConfAndExec() {
            try {
                if (sentinelConf == null) {
                    resolveSentinelConf();
                }
                executable = redisExecProvider.get();
            } catch (Exception e) {
                throw new RedisBuildingException("Could not build sentinel instance", e);
            }
        }

        public void reset() {
            this.redisConfigBuilder = null;
            this.sentinelConf = null;
        }

        public void addDefaultReplicationGroup() {
            setting(String.format(MASTER_MONITOR_LINE, masterName, masterPort, quorumSize));
            setting(String.format(DOWN_AFTER_LINE, masterName, downAfterMilliseconds));
            setting(String.format(FAIL_OVER_LINE, masterName, failOverTimeout));
            setting(String.format(PARALLEL_SYNCS_LINE, masterName, parallelSyncs));
        }

        private void resolveSentinelConf() throws IOException {
            if (redisConfigBuilder == null) {
                addDefaultReplicationGroup();
            }
            setting(String.format(PORT_LINE, port));
            final String configString = redisConfigBuilder.toString();

            File redisConfigFile = File.createTempFile(resolveConfigName(), ".conf");
            redisConfigFile.deleteOnExit();
            Files.write(configString, redisConfigFile, Charset.forName("UTF-8"));
            sentinelConf = redisConfigFile.getAbsolutePath();
        }

        private String resolveConfigName() {
            return CONF_FILENAME + "_" + port;
        }

        private List<String> buildCommandArgs() {
            Preconditions.checkNotNull(sentinelConf);

            List<String> args = new ArrayList<>();
            args.add(executable.getAbsolutePath());
            args.add(sentinelConf);
            args.add("--sentinel");

            if (port != null) {
                args.add("--port");
                args.add(Integer.toString(port));
            }

            return args;
        }
    }
}


src/main/java/redis/embedded/RedisExecProvider.java
package redis.embedded;

import com.google.common.base.Preconditions;
import com.google.common.collect.Maps;
import redis.embedded.util.*;

import java.io.File;
import java.io.IOException;
import java.util.Map;
import java.util.regex.Pattern;

public class RedisExecProvider {

    private final Map<OsArchitecture, RedisExecutable> executables = Maps.newHashMap();
    public final static Pattern DEFAULT_REDIS_READY_PATTERN = Pattern.compile(
            "(?:The server is now ready to accept connections on port)" +   // 3.2.1, 2.8.24
                    "|(?:Ready to accept connections)"  + // 4.0.2
                    "|(?:Sentinel ID is)" +  // 3.2.1, 4.0.2
                    "|(?:Sentinel runid is)" // 2.8.24
    );

    /**
     * @return a new RedisExecProvider instance
     * @deprecated use {@link #build()} instead
     */
    @Deprecated
    public static RedisExecProvider defaultProvider() {
        return new RedisExecProvider();
    }

    public static RedisExecProvider build() {
        return new RedisExecProvider();
    }

    private RedisExecProvider() {
        initExecutables();
    }

    private void initExecutables() {
        executables.put(OsArchitecture.UNIX_x86_64, RedisExecutable.build("redis-server-3.0.7", DEFAULT_REDIS_READY_PATTERN));

        executables.put(OsArchitecture.MAC_OS_X_x86, RedisExecutable.build("redis-server-3.0.7-darwin", DEFAULT_REDIS_READY_PATTERN));
        executables.put(OsArchitecture.MAC_OS_X_x86_64, RedisExecutable.build("redis-server-3.0.7-darwin", DEFAULT_REDIS_READY_PATTERN));
    }

    public RedisExecProvider override(OS os, String executable) {
        Preconditions.checkNotNull(executable);
        for (Architecture arch : Architecture.values()) {
            override(os, arch, executable);
        }
        return this;
    }

    public RedisExecProvider override(OS os, Architecture arch, String executable) {
        Preconditions.checkNotNull(executable);
        executables.put(new OsArchitecture(os, arch), RedisExecutable.build(executable, DEFAULT_REDIS_READY_PATTERN));
        return this;
    }

    public RedisExecProvider override(OS os, Architecture arch, RedisExecutable redisExec) {
        Preconditions.checkNotNull(redisExec);
        executables.put(new OsArchitecture(os, arch), redisExec);
        return this;
    }

    public File get() throws IOException {
        OsArchitecture osArch = OsArchitecture.detect();
        String executablePath = executables.get(osArch).getExecutableName();
        return fileExists(executablePath) ?
                new File(executablePath) :
                JarUtil.extractExecutableFromJar(executablePath);

    }

    public Pattern getExecutableStartPattern() {
        OsArchitecture osArch = OsArchitecture.detect();
        return executables.get(osArch).getStartPattern();
    }

    public RedisExecProvider copy() {
        RedisExecProvider copy = new RedisExecProvider();

        for (OsArchitecture k : executables.keySet()) {
            copy.override(k.os(), k.arch(), executables.get(k));
        }

        return copy;
    }

    private boolean fileExists(String executablePath) {
        return new File(executablePath).exists();
    }
}


src/main/java/redis/embedded/util/RedisExecutor.java
package redis.embedded.util;

import java.util.regex.Pattern;

public class RedisExecutor {
    private String executableName;
    private Pattern startPattern;

    public RedisExecutor(String exeName, Pattern startPattern) {
        this.executableName = exeName;
        this.startPattern = startPattern;
    }

    public String getExecutableName() {
        return executableName;
    }

    public Pattern getStartPattern() {
        return startPattern;
    }

    public String getStartPatternAsString() {
        return startPattern.pattern();
    }

    public RedisExecutor copy() {
        return new RedisExecutor(executableName, startPattern);
    }

    @Override
    public boolean equals(Object obj) {
        return obj instanceof RedisExecutor && ((RedisExecutor) obj).executableName.contentEquals(executableName) &&
                ((RedisExecutor)obj).startPattern.pattern().contentEquals(startPattern.pattern());
    }

    @Override
    public int hashCode() {
        return executableName.hashCode() + startPattern.pattern().hashCode();
    }
}


src/main/java/redis/embedded/util/RedisExecutable.java
package redis.embedded.util;

import java.util.regex.Pattern;

public class RedisExecutable {
    private String executableName;
    private Pattern startPattern;

    public static RedisExecutable build(String execName, Pattern startPattern) {
        return new RedisExecutable(execName, startPattern);
    }

    private RedisExecutable(String exeName, Pattern startPattern) {
        this.executableName = exeName;
        this.startPattern = startPattern;
    }

    public String getExecutableName() {
        return executableName;
    }

    public Pattern getStartPattern() {
        return startPattern;
    }

    public String getStartPatternAsString() {
        return startPattern.pattern();
    }

    public RedisExecutable copy() {
        return new RedisExecutable(executableName, startPattern);
    }

    @Override
    public boolean equals(Object obj) {
        return obj instanceof RedisExecutable && ((RedisExecutable) obj).executableName.contentEquals(executableName) &&
                ((RedisExecutable)obj).startPattern.pattern().contentEquals(startPattern.pattern());
    }

    @Override
    public int hashCode() {
        return executableName.hashCode() + startPattern.pattern().hashCode();
    }
}


src/main/java/redis/embedded/AbstractRedisInstance.java
package redis.embedded;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import redis.embedded.cluster.RedisCluster;
import redis.embedded.exceptions.EmbeddedRedisException;

import java.io.*;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;
import java.util.regex.Pattern;

import static org.apache.commons.io.IOUtils.closeQuietly;

abstract class AbstractRedisInstance implements Redis {

    private static final Logger logger = LoggerFactory.getLogger(RedisCluster.class);

    private static final long EXECUTOR_SHUTDOWN_TIMEOUT_MS = 10000;

    List<String> args = Collections.emptyList();
    private State state = new State();
    private Process redisProcess;
    private final int port;
    private boolean logProcessOutput;
    private long startupTimeoutMs = 3000;

    private LogsInputStream logsInput = new LogsInputStream();

    AbstractRedisInstance(int port) {
        this.port = port;
    }

    @Override
    public boolean isActive() {
        return state.isActive();
    }

    @Override
    public void start() throws EmbeddedRedisException {
        if (state.setActivating()) {
            try {
                redisProcess = createRedisProcessBuilder().start();

                awaitRedisServerReady();
            } catch (IOException | InterruptedException e) {
                throw new EmbeddedRedisException("Failed to start Redis instance", e);
            }
        } else {
            throw new EmbeddedRedisException("This redis server instance is already running...");
        }
    }

    @Override
    public InputStream errors() {
        return logsInput;
    }

    private void awaitRedisServerReady() throws IOException, InterruptedException {
        if (!state.isActivating()) {
            throw new RuntimeException("Internal error. State is not activating in await: " + state);
        }
        
        BufferedReader reader = new BufferedReader(new InputStreamReader(redisProcess.getInputStream()));
        ScheduledExecutorService executor = setupStartupTimeoutCheck();
        try {
            // Wait 200ms to be sure redis have started (it can takes some time and that would cause
            // the exception below to throw.
            Thread.sleep(200);

            String outputLine;
            while (true) {
                outputLine = reader.readLine();
                if (outputLine == null) {
                    //Something goes wrong. Stream is ended before server was activated.
                    //Set state to failed
                    if (state.setFailed()) {
                        logger.error("Invalid state during wait: {}", state);
                        break;
                    } else {
                        throw new RuntimeException("Can't start redis server. Check logs for details.");
                    }
                }
                logsInput.appendLine(outputLine);
                if (logProcessOutput) {
                    logger.info(outputLine);
                }
                if (redisReadyPattern().matcher(outputLine).find()) {
                    if (!state.setActive()) {
                        throw new RuntimeException("Redis server was deactivated before receiving ready pattern log");
                    }
                    break;
                }
            }
        } finally {
            closeQuietly(reader);
            cancelStartupTimeoutCheck(executor);
        }
    }

    private ScheduledExecutorService setupStartupTimeoutCheck() {
        ScheduledExecutorService scheduledExecutorService = Executors.newSingleThreadScheduledExecutor();
        scheduledExecutorService.schedule(new Runnable() {
            @Override
            public void run() {
                if (state.setDeactivatingByTimeout()) {
                    doStop();
                }
            }
        }, startupTimeoutMs, TimeUnit.MILLISECONDS);
        return scheduledExecutorService;
    }

    private void cancelStartupTimeoutCheck(ScheduledExecutorService executor) {
        if (executor != null) {
            executor.shutdownNow();
            try {
                executor.awaitTermination(EXECUTOR_SHUTDOWN_TIMEOUT_MS, TimeUnit.MILLISECONDS);
            } catch (InterruptedException e) {
                logger.warn("Error waiting for startup timeout check executor termination", e);
            }
        }
    }

    protected abstract Pattern redisReadyPattern();

    private ProcessBuilder createRedisProcessBuilder() {
        File executable = new File(args.get(0));
        ProcessBuilder pb = new ProcessBuilder(args)
                .directory(executable.getParentFile())
                .redirectErrorStream(true);
        return pb;
    }

    @Override
    public void stop() throws EmbeddedRedisException {
        if (state.setDeactivating()) {
            doStop();
            if (!state.setInactive()) {
                throw new RuntimeException("Internal error. Can't set redis state to inactiveState from " + state);
            }
        } else {
            logger.debug("Attempt to stop reddis while in {}", state);
        }
    }

    private void doStop() throws EmbeddedRedisException {
        if (!state.isDeactivating()) {
            throw new RuntimeException("Internal error. Can't stop redis because it's not in deactivating state. But " + state);
        }
        redisProcess.destroy();
        tryWaitFor();
    }

    private void tryWaitFor() {
        try {
            redisProcess.waitFor();
        } catch (InterruptedException e) {
            throw new EmbeddedRedisException("Failed to stop redis instance", e);
        }
    }

    @Override
    public List<Integer> ports() {
        return Collections.singletonList(port);
    }

    void setLogProcessOutput(boolean logProcessOutput) {
        this.logProcessOutput = logProcessOutput;
    }

    void setStartupTimeoutMs(long startupTimeoutMs) {
        this.startupTimeoutMs = startupTimeoutMs;
    }

    RedisState getState() {
        return state.getState();
    }
}


src/main/java/redis/embedded/RedisServer.java
package redis.embedded;

import com.google.common.base.Preconditions;
import com.google.common.base.Strings;
import com.google.common.io.Files;
import redis.embedded.exceptions.RedisBuildingException;

import java.io.File;
import java.io.IOException;
import java.net.InetSocketAddress;
import java.nio.charset.Charset;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.regex.Pattern;

public class RedisServer extends AbstractRedisInstance {

    private Pattern redisReadyPat = RedisExecProvider.DEFAULT_REDIS_READY_PATTERN;

    private static final int DEFAULT_REDIS_PORT = 6379;

    public RedisServer() throws IOException {
        this(DEFAULT_REDIS_PORT);
    }

    public RedisServer(int port) throws IOException {
        super(port);
        File executable = RedisExecProvider.build().get();
        this.args = Arrays.asList(
                executable.getAbsolutePath(),
                "--port", Integer.toString(port)
        );
    }

    public RedisServer(RedisExecProvider redisExecProvider, int port) throws IOException {
        super(port);
        this.redisReadyPat = redisExecProvider.getExecutableStartPattern();
        this.args = Arrays.asList(
                redisExecProvider.get().getAbsolutePath(),
                "--port", Integer.toString(port)
        );
    }

    RedisServer(List<String> args, int port) {
        super(port);
        this.args = new ArrayList<>(args);
    }

    void setRedisReadyPat(Pattern redisReadyPat) {
        Preconditions.checkNotNull(redisReadyPat);
        this.redisReadyPat = redisReadyPat;
    }

    /**
     * Prefer using {@code new RedisServer.Builder()} directly.
     *
     * @return a new Builder for this class
     */
    @Deprecated
    public static Builder builder() {
        return new Builder();
    }

    @Override
    protected Pattern redisReadyPattern() {
        return redisReadyPat;
    }

    @SuppressWarnings("unused")
    public static class Builder {
        private static final String LINE_SEPARATOR = System.getProperty("line.separator");
        private static final String CONF_FILENAME = "embedded-redis-server";
        private static final long DEFAULT_STARTUP_TIMEOUT_MS = 3000;

        private File executable;
        private RedisExecProvider redisExecProvider = RedisExecProvider.build();
        private int port = 6379;
        private InetSocketAddress slaveOf;

        private String redisConf;
        private StringBuilder redisConfigBuilder;
        private boolean logProcessOutput;
        private long startupTimeoutMs = DEFAULT_STARTUP_TIMEOUT_MS;

        public Builder redisExecProvider(RedisExecProvider redisExecProvider) {
            this.redisExecProvider = redisExecProvider;
            return this;
        }

        public Builder port(int port) {
            this.port = port;
            return this;
        }

        public Builder slaveOf(String hostname, int port) {
            this.slaveOf = new InetSocketAddress(hostname, port);
            return this;
        }

        public Builder slaveOf(InetSocketAddress slaveOf) {
            this.slaveOf = slaveOf;
            return this;
        }

        public Builder configFile(String redisConf) {
            if (redisConfigBuilder != null) {
                throw new RedisBuildingException("Redis configuration is already partially build using setting(String) method!");
            }
            this.redisConf = redisConf;
            return this;
        }

        public Builder logProcessOutput() {
            this.logProcessOutput = true;
            return this;
        }

        public Builder startupTimeoutMs(long startupTimeoutMs) {
            this.startupTimeoutMs = startupTimeoutMs;
            return this;
        }

        @SuppressWarnings("Duplicates")
        public Builder setting(String configLine) {
            if (redisConf != null) {
                throw new RedisBuildingException("Redis configuration is already set using redis conf file!");
            }

            if (redisConfigBuilder == null) {
                redisConfigBuilder = new StringBuilder();
            }

            redisConfigBuilder.append(configLine);
            redisConfigBuilder.append(LINE_SEPARATOR);
            return this;
        }

        public RedisServer build() {
            tryResolveConfAndExec();
            List<String> args = buildCommandArgs();
            RedisServer redisServer = new RedisServer(args, port);
            redisServer.setLogProcessOutput(logProcessOutput);
            redisServer.setStartupTimeoutMs(startupTimeoutMs);
            redisServer.setRedisReadyPat(redisExecProvider.getExecutableStartPattern());
            return redisServer;
        }

        public void reset() {
            this.executable = null;
            this.redisConfigBuilder = null;
            this.slaveOf = null;
            this.redisConf = null;
        }

        public RedisServer.Builder copy() {
            Builder newBuilder = new Builder();

            newBuilder.executable = (executable != null)
                    ? new File(executable.toURI())
                    : null;
            newBuilder.redisExecProvider = redisExecProvider.copy();
            newBuilder.port = port;
            newBuilder.slaveOf = (slaveOf != null)
                    ? new InetSocketAddress(slaveOf.getAddress(), slaveOf.getPort())
                    : null;
            newBuilder.redisConf = redisConf;
            newBuilder.redisConfigBuilder = (redisConfigBuilder != null)
                    ? new StringBuilder(redisConfigBuilder)
                    : null;
            newBuilder.logProcessOutput = logProcessOutput;

            return newBuilder;
        }

        /**
         * @return the configuration built using {@code #setting(String)}.
         */
        public String settings() {
            return redisConfigBuilder.toString();
        }

        private void tryResolveConfAndExec() {
            try {
                resolveConfAndExec();
            } catch (IOException e) {
                throw new RedisBuildingException("Could not build server instance", e);
            }
        }

        private void resolveConfAndExec() throws IOException {
            if (redisConf == null && redisConfigBuilder != null) {
                File redisConfigFile = File.createTempFile(resolveConfigName(), ".conf");
                redisConfigFile.deleteOnExit();
                Files.write(redisConfigBuilder.toString(), redisConfigFile, Charset.forName("UTF-8"));
                redisConf = redisConfigFile.getAbsolutePath();
            }

            try {
                executable = redisExecProvider.get();
            } catch (Exception e) {
                throw new RedisBuildingException("Failed to resolve executable", e);
            }
        }

        private String resolveConfigName() {
            return CONF_FILENAME + "_" + port;
        }

        private List<String> buildCommandArgs() {
            List<String> args = new ArrayList<>();
            args.add(executable.getAbsolutePath());

            if (!Strings.isNullOrEmpty(redisConf)) {
                args.add(redisConf);
            }

            args.add("--port");
            args.add(Integer.toString(port));

            if (slaveOf != null) {
                args.add("--slaveof");
                args.add(slaveOf.getHostName());
                args.add(Integer.toString(slaveOf.getPort()));
            }

            return args;
        }
    }
}
